hudi-aws/src/main/java/org/apache/hudi/aws/transaction/lock/S3StorageLockClient.java

```diff
@@ -0,0 +1,248 @@
+   * Internal helper to create or update the lock file with optional ETag precondition.
+   */
+  private StorageLockFile createOrUpdateLockFileInternal(StorageLockData lockData, String expectedEtag) {
+    byte[] bytes = StorageLockFile.toByteArray(lockData);
+    PutObjectRequest.Builder putRequestBuilder = PutObjectRequest.builder()
+            .bucket(bucketName)
+            .key(lockFilePath);
+
+    // ETag-based constraints:
+    // - If expectedEtag is not null:
+    //    We assume that the file already exists on S3 with the ETag "expectedEtag".
+    //    The update operation will include an ifMatch(expectedEtag) condition, meaning the update will only
+    //    succeed if the current file's ETag exactly matches expectedEtag.
+    //    If the actual ETag of the file on S3 differs from expectedEtag, the update attempt will fail.
+    // - If expectedEtag is null:
+    //    We assume that the file does not currently exist on S3.
+    //    The operation will use ifNoneMatch("*"), which instructs S3 to create the file only if it doesn't already exist.
+    //    If a file with the same name is present (i.e., there is an existing ETag), the creation attempt will fail.
+    if (expectedEtag == null) {
+      putRequestBuilder.ifNoneMatch("*");
+    } else {
+      putRequestBuilder.ifMatch(expectedEtag);
+    }
+
+    PutObjectResponse response = s3Client.putObject(putRequestBuilder.build(), RequestBody.fromBytes(bytes));
+    String newEtag = response.eTag();
+
+    return new StorageLockFile(lockData, newEtag);
+  }
+
+  @Override
+  public Pair<LockUpdateResult, StorageLockFile> tryCreateOrUpdateLockFile(
+          StorageLockData newLockData,
+          StorageLockFile previousLockFile
+  ) {
+    String currentEtag = (previousLockFile != null) ? previousLockFile.getVersionId() : null;
+    try {
+      StorageLockFile updatedFile = createOrUpdateLockFileInternal(newLockData, currentEtag);
+      return Pair.of(LockUpdateResult.SUCCESS, updatedFile);
+    } catch (S3Exception e) {
+      int status = e.statusCode();
+      if (status == PRECONDITION_FAILURE_ERROR_CODE || status == CONDITIONAL_REQUEST_CONFLICT_ERROR_CODE) {
+        logger.info("OwnerId: {}, Lockfile modified by another process: {}", ownerId, lockFilePath);
+        return Pair.of(LockUpdateResult.ACQUIRED_BY_OTHERS, null);
+      } else if (status == RATE_LIMIT_ERROR_CODE) {
+        logger.warn("OwnerId: {}, Rate limit exceeded for: {}", ownerId, lockFilePath);
+      } else if (status >= INTERNAL_SERVER_ERROR_CODE_MIN) {
+        logger.warn("OwnerId: {}, S3 internal server error for: {}", ownerId, lockFilePath, e);
+      } else {
+        throw e;
+      }
+      return Pair.of(LockUpdateResult.UNKNOWN_ERROR, null);
+    }
+  }
+
+  @Override
+  public Pair<LockUpdateResult, StorageLockFile> tryCreateOrUpdateLockFileWithRetry(
+          Supplier<StorageLockData> newLockDataSupplier,
+          StorageLockFile previousLockFile,
+          long maxAttempts
```

Comment:
same. Check if this needs to be tied to `LockConfiguration#LOCK_ACQUIRE_NUM_RETRIES_PROP_KEY` in some way?